clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3047,7 +3062,7 @@ class SYCLKernelNameTypeVisitor
           IsInvalid = true;
           return;
         }
-        if (Tag->isCompleteDefinition()) {
+        if (isa<FunctionDecl>(DeclCtx) && Tag->isCompleteDefinition()) {
```

Comment:
The if at L3065 is to handle cases like the following:
```
       void test() {
        class InvalidKernelName1 {};
          q.submit([&](handler &h) {
            h.single_task<InvalidKernelName1>([]() {});
          });
```
The order of decl context for  InvalidKernelName1 is 
**clang::Decl::Function**
clang::Decl::TranslationUnit

And  `class InvalidKernelName1 {};` is completely defined.

Reply:
I guess I'm more asking for a situation where that isa is required.  It seems that both your examples are positive things here, but what cases are you trying to exclude from these 'if's?

Reply:
Trying to exclude cases like nested named structs and classes inside structs.

Reply:
I unfortunately don't know the visitor here well enough to know exactly what is going on here, but this part of the change "feels" wrong, but I obviously cannot specify what makes me think that.  It just generally seems like the visitor isn't doing what it is supposed to if all this is necessary, and these two 'if' statements are particularly constrained for situations that are not sufficiently thought out.

I'm hoping that @Fznamznon can analyze the patch/problem and make sure this is the right approach.